### PR TITLE
Some improvements to k-buckets

### PR DESCRIPTION
**Pull request by tomaka**
The idea is that when writing your own topology, you should be able to use the `KBuckets` struct for storage.

- Publicly reexport the `kbuckets` module.
- Remove interior mutability.
- Simplify the `KBucketsPeerId` trait.

Ultimately Kademlia would also implement being a record store, but unfortunately I don't really know how to plug that together yet.

### CODE REVIEW

File: protocols/kad/src/kbucket.rs [hunk 1]
```diff
@@ -30,65 +30,53 @@
...
-#[derive(Debug)]
+/// Table of k-buckets.
+#[derive(Debug, Clone)]
 pub struct KBucketsTable<Id, Val> {
+    /// Peer ID of the local node.
     my_id: Id,
-    tables: Vec<Mutex<KBucket<Id, Val>>>,
+    /// The actual tables.
+    tables: Vec<KBucket<Id, Val>>,
     // The timeout when pinging the first node after which we consider that it no longer responds.
```
Comment (dvdplm) [inline protocols/kad/src/kbucket.rs]:
```suggestion
     // The timeout when pinging the first node after which we consider it unresponsive.
```

File: protocols/kad/src/kbucket.rs [hunk 2]
```diff
@@ -30,65 +30,53 @@
...
 
-/// Table of k-buckets with interior mutability.
-#[derive(Debug)]
+/// Table of k-buckets.
+#[derive(Debug, Clone)]
 pub struct KBucketsTable<Id, Val> {
+    /// Peer ID of the local node.
     my_id: Id,
-    tables: Vec<Mutex<KBucket<Id, Val>>>,
+    /// The actual tables.
```
Comment (dvdplm) [inline protocols/kad/src/kbucket.rs]:
```suggestion
     /// The actual tables that store peers or values.
```
maybe?

File: protocols/kad/src/kbucket.rs [hunk 3]
```diff
@@ -30,65 +30,53 @@
...
+    /// Nodes are always ordered from oldest to newest.
+    /// Note that we will very often move elements to the end of this. No benchmarking has been
+    /// performed, but it is very likely that a `ArrayVec` is the most performant data structure.
     nodes: ArrayVec<[Node<Id, Val>; MAX_NODES_PER_BUCKET]>,
 
-    // Node received when the bucket was full. Will be added to the list if the first node doesn't
-    // respond in time to our ping. The second element is the time when the pending node was added.
-    // If it is too much in the past, then we drop the first node and add the pending node to the
-    // end of the list.
+    /// Node received when the bucket was full. Will be added to the list if the first node doesn't
```
Comment (dvdplm) [inline protocols/kad/src/kbucket.rs]:
```suggestion
     /// Nodes received when the bucket was full will be added to the list if the first node doesn't
```

File: protocols/kad/src/kbucket.rs [hunk 4]
```diff
@@ -30,65 +30,53 @@
...
+    /// respond in time to our ping. The second element is the time when the pending node was added.
+    /// If it is too much in the past, then we drop the first node and add the pending node to the
```
Comment (dvdplm) [inline protocols/kad/src/kbucket.rs]:
```suggestion
     /// If it is too old we drop the first node and add the pending node to the
```

File: protocols/kad/src/kbucket.rs [hunk 5]
```diff
@@ -30,65 +30,53 @@
...
+    value: Val,
+}
+
 impl<Id, Val> KBucket<Id, Val> {
-    // Puts the kbucket into a coherent state.
-    // If a node is pending and the timeout has expired, removes the first element of `nodes`
-    // and pushes back the node in `pending_node`.
+    /// Puts the kbucket into a coherent state.
+    /// If a node is pending and the timeout has expired, removes the first element of `nodes`
+    /// and pushes back the node in `pending_node`.
```
Comment (dvdplm) [inline protocols/kad/src/kbucket.rs]:
```suggestion
     /// and puts the node back in `pending_node`.
```

(highly subjective, but to me "pushes" sounds a little like there could be more than one item pending; feel free to ignore if you disagree!)

File: protocols/kad/src/kbucket.rs (hunk 3 above)
Comment (tomaka) [reply protocols/kad/src/kbucket.rs]: It's only a single node. What this comment says is that the content of `pending_node` can be added to `nodes` at a later point in time.

**Pull request merged**